javalib/src/main/scala/java/util/concurrent/CopyOnWriteArrayList.scala

```diff
@@ -0,0 +1,348 @@
+package java.util
+package concurrent
+
+import java.lang.{reflect => jlr}
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+
+import scala.scalajs._
+
+class CopyOnWriteArrayList[E](private var inner: js.Array[E])
+    extends AbstractCopyOnWriteArrayList[E] with List[E] with RandomAccess
+    with Cloneable with Serializable { self =>
+
+  // requiresCopyOnWrite is false if and only if no other object (like the iterator)
+  // may have a reference to inner
+  private var requiresCopyOnWrite = false
+
+  def this() = {
+    this(new js.Array[E])
+  }
+
+  def this(c: Collection[_ <: E]) = {
+    this()
+    addAll(c)
+  }
+
+  def this(toCopyIn: Array[E]) = {
+    this()
+    toCopyIn.foreach(add)
+  }
+
+  def size(): Int =
+    inner.size
+
+  override def clone(): AnyRef =
+    new CopyOnWriteArrayList[E](this)
+
+  def clear(): Unit = {
+    inner = new js.Array[E]
+    requiresCopyOnWrite = false
+  }
+
+  def listIterator(index: Int): ListIterator[E] = {
+    checkIndexOnBounds(index)
+    new CopyOnWriteArrayListIterator[E](innerSnapshot(), index, 0, size)
+  }
+
+  def subList(fromIndex: Int, toIndex: Int): List[E] = {
+    if (fromIndex < 0 || fromIndex > toIndex || toIndex > size )
+      throw new  IndexOutOfBoundsException
+    new CopyOnWriteArrayListView(fromIndex, toIndex)
+  }
+
+  protected def innerGet(index: Int): E =
+    inner(index)
+
+  protected def innerSet(index: Int, elem: E): Unit =
+    inner(index) = elem
```

Comment:
Wouldn't it make more sense to move `copyIfNeeded()` from `set(index, elem)` to `innerSet()`? `copyIfNeeded()` looks more of an implementation detail related to `innerSet` than `set`, because `copyIfNeeded()` is manipulating `inner`, at the same level of abstraction as `innerSet`. The same reasoning applies to other `innerMutate` methods.